File: src/main/java/storage/TestQuestion.java
/*
 * Copyright 2020 Oliver Trevor and Suchin Ravi.
 *
 * This file is part of MarkSpace.
 *
 * MarkSpace is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * MarkSpace is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MarkSpace.  If not, see <https://www.gnu.org/licenses/>.
 */

package storage;

public class TestQuestion {
    private int test_question_id;
    private int page;
    private String points;
    private CanvasRectangle[] regions;
    private boolean extra_credit;

    public int getTestQuestionId() {
        return test_question_id;
    }

    public void setTestQuestionId(int test_question_id) {
        this.test_question_id = test_question_id;
    }

    public int getPage() {
        return page;
    }

    public String getPoints() {
        return points;
    }

    public boolean getExtraCredit() {
        return extra_credit;
    }

    public void setExtraCredit(boolean extra_credit) {
        this.extra_credit = extra_credit;
    }

    public CanvasRectangle[] getRegions() {
        return regions;
    }

    public void setPage(int page) {
        this.page = page;
    }

    public void setPoints(String points) {
        this.points = points;
    }

    public void setRegions(CanvasRectangle[] regions) {
        this.regions = regions;
    }

    public boolean isIdentificationQuestion() {
        return this.regions[0].getLabel().equals("NAME");
    }

    public static class CanvasRectangle {
        private double x;
        private double y;
        private int layer;
        private double width;
        private double height;
        private String color;
        private String outline_color;
        private String label;

        public double getX() {
            return x;
        }

        public void setX(double x) {
            this.x = x;
        }

        public double getY() {
            return y;
        }

        public void setY(double y) {
            this.y = y;
        }

        public int getLayer() {
            return layer;
        }

        public void setLayer(int layer) {
            this.layer = layer;
        }

        public double getWidth() {
            return width;
        }

        public void setWidth(double width) {
            this.width = width;
        }

        public double getHeight() {
            return height;
        }

        public void setHeight(double height) {
            this.height = height;
        }

        public String getColor() {
            return color;
        }

        public void setColor(String color) {
            this.color = color;
        }

        public String getOutlineColor() {
            return outline_color;
        }

        public void setOutlineColor(String outline_color) {
            this.outline_color = outline_color;
        }

        public String getLabel() {
            return label;
        }

        public void setLabel(String label) {
            this.label = label;
        }
    }
}

File: src/main/resources/static/scripts/typescript/testquestion.ts
class TestQuestion {
    private page: number;
    private points: string;
    private regions: CanvasRectangle[];
    private extra_credit: boolean;

    constructor(points: string, page: number, regions: CanvasRectangle[], extra_credit: boolean) {
        this.points = points;
        this.page = page;
        this.regions = regions;
        this.extra_credit = extra_credit;
    }

    equals(other: TestQuestion): boolean {
        if (other.getRegions().length != this.regions.length) {
            return false;
        }
        for (let i = 0; i < other.getRegions().length; i++) {
            if (!other.getRegions()[i].equals(this.regions[i])) {
                return false;
            }
        }
        return other.page == this.page && other.points == this.points && other.extra_credit == this.extra_credit;
    }

    getPoints(): string {
        return this.points;
    }

    getPage(): number {
        return this.page;
    }

    setPoints(points: string): void {
        this.points = points;
    }

    getRegions(): CanvasRectangle[] {
        return this.regions;
    }

    setRegions(regions: CanvasRectangle[]): void {
        this.regions = regions;
    }

    getExtraCredit(): boolean {
        return this.extra_credit;
    }

    setExtraCredit(extra_credit: boolean): void {
        this.extra_credit = extra_credit;
    }

    getImageURLs(test_id: number, answers: boolean): string[] {
        const image_urls: string[] = new Array<string>();
        for (let region of this.regions) {
            const image_url: string = `/render_question?page=${this.page}&test_id=${test_id}&answers=${answers}&x=${region.getX()}&y=${region.getY()}&width=${region.getWidth()}&height=${region.getHeight()}&dpi=${page_render_dpi}`;
            image_urls.push(image_url);
        }
        return image_urls;
    }

    static fromRawObject(test_question_raw: object): TestQuestion {
        const regions: CanvasRectangle[] = new Array<CanvasRectangle>();
        // @ts-ignore
        for (let region_raw of test_question_raw.regions) {
            const region: CanvasRectangle = new CanvasRectangle(region_raw.x, region_raw.y, region_raw.width, region_raw.height, region_raw.color, region_raw.outline_color, region_raw.label, region_raw.layer);
            regions.push(region);
        }
        // @ts-ignore
        const test_question: TestQuestion = new TestQuestion(test_question_raw.points, test_question_raw.page, regions, test_question_raw.extra_credit);
        return test_question;
    }
}
